### ISSUE TITLE
Javaslatok I.

### ISSUE THREAD

**Issue by qbeer** [szmate00/elte_advanced_ml_lab#1]
Szia,

 kicsit egyszerűbb lenne ha struktúrálnád a kódot, én megrögzötten utálom a jupyter notebookot, de most inkább arra gondolok, hogy érdemes lenne az adatgenerálást egy külön fájlba tenni, állíítható paraméterekkel.

* pl. most szerintem nagyon rövid time-span-en tanítod a modellt, alig néhány pont van a p-q diagrammon, szóval keveset lép a rednszered szerintem nagyon

Az adatgenerálás a rugós rendszerre jónak tűnik, de szerintem azt beszéltük, hogy lesz még egy Hamiltoni amit lehet majd használni itt az első körben, egyenlőre szerintem koncentráljunk arra, hogy az adatgenerálás szépen meg legyen. 👍🏻 Ugye ez azért  is fontos, mert nem biztos, hogy az csak 1 dimenziós rendszer lesz, mint ez.

--------------------------

### Tanító kód

A sztenderd MLP jól néz ki, meg a traning loop is rendben van, tanul is, egyenlőre ez így még nem konklúzív, kellenek majd nyilván a prediktált trajektóriák a teszt halmahzon, azokon kiértékelve a loss és a q-p diagrammok.

A Hamiltoni hálónál szintén szerintem nagyjából rendben van a dolog, de én ezt a részt beépíteném a modulba:

```python
# differentiating the output of the backbone HNN
# and returning appropiate time derivates of inputs
def diff_hnn(X):
    X = torch.autograd.Variable(X, requires_grad=True)
    out = torch.autograd.grad(model(X).sum(), X, create_graph=True)[0]
    
    return out @ torch.Tensor([[0, -1],[1, 0]]) # dHdq, dHdp -> dqdt, dpdt
```

Ezt be tudnád építeni a forward passba, nem vagyok benne biztos, hogy ez így pontosan azt optimalizálja amit te szeretnél. Könnyebb nem kidebuggolni, mint jobban átgondolni a modellt.

Ez itt számszerűen jó, de nyilván úgy tudod jól összehasonlítani a modelleket, ha tanítás közben kiírod a MSE loss-t a dy, dy_hat között mindkét modellnél.

A kiértékelésnél ugye ez lenne a cél, mint amikor legenerálod az adatot, hogy a basline modellből kinyered egyből a deriváltakat és a hnn modellnél is ez a cél, csak ott egy olyan függvényt kell írnod, ami H alapján visszaadja azokat:

```python
symplectic_ivp = solve_ivp(fun=<ide-jön-majd-a-neurális-háló-ami-deriváltakat-ad-vissza>, t_span=[0, 20], y0=y0, t_eval=T)
```

Alex
